tests/phpunit/integration/Core/Authentication/AuthenticationTest.php

```diff
@@ -235,6 +242,134 @@ public function test_register_allowed_redirect_hosts() {
 		$this->assertEquals( 'https://sitekit.withgoogle.com', wp_validate_redirect( 'https://sitekit.withgoogle.com' ) );
 	}
 
+	public function test_register_maybe_refresh_token_for_screen__admin() {
+		$user_id = $this->factory()->user->create( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user_id );
+		$context      = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$user_options = new User_Options( $context, $user_id );
+
+		$auth = new Authentication( $context, null, $user_options );
+
+		remove_all_actions( 'current_screen' );
+		remove_all_actions( 'heartbeat_tick' );
+		$auth->register();
+		$this->assertTrue( has_action( 'current_screen' ) );
+		$this->assertTrue( has_action( 'heartbeat_tick' ) );
+
+		$oauth_client = $auth->get_oauth_client();
+		// Fake a valid authentication token on the oauth client.
+		$this->assertTrue(
+			$oauth_client->set_token(
+				array(
+					'access_token'  => 'test-access-token',
+					'refresh_token' => 'test-refresh-token',
+				)
+			)
+		);
+		// The FakeHttpClient returns 200 by default.
+		$oauth_client->get_client()->setHttpClient( new FakeHttpClient() );
+		// Make sure we start with no errors.
+		$this->assertFalse( get_user_option( OAuth_Client::OPTION_ERROR_CODE, $user_id ) );
+
+		// Token should not refresh on any screen other other than the dashboard.
+		do_action( 'current_screen', WP_Screen::get( 'some-random-screen' ) );
+		// There should be no errors as refresh_token() should not be called as yet.
+		$this->assertFalse( get_user_option( OAuth_Client::OPTION_ERROR_CODE, $user_id ) );
+
+		// Token should not refresh if a user does not have credentials.
+		do_action( 'current_screen', WP_Screen::get( 'toplevel_page_googlesitekit-dashboard' ) );
+		// There should be no errors as refresh_token() should not be called as yet.
+		$this->assertFalse( get_user_option( OAuth_Client::OPTION_ERROR_CODE, $user_id ) );
+
+		// Emulate credentials.
+		$this->fake_proxy_site_connection();
+
+		// Token should still not refresh as it expires after 5 minutes.
+		do_action( 'current_screen', WP_Screen::get( 'toplevel_page_googlesitekit-dashboard' ) );
+		// There should be no errors as refresh_token() should not be called as yet.
+		$this->assertFalse( get_user_option( OAuth_Client::OPTION_ERROR_CODE, $user_id ) );
+
+		// Set the user's token to expire within 5 minutes.
+		$user_options->set( OAuth_Client::OPTION_ACCESS_TOKEN_EXPIRES_IN, 295 );
+		// Token should refresh now as all conditions have been met.
+		do_action( 'current_screen', WP_Screen::get( 'toplevel_page_googlesitekit-dashboard' ) );
+		// There is no actual response when refresh_token() is executed - so this is a successful call to refresh_token().
+		$this->assertEquals( 'Invalid JSON response', get_user_option( OAuth_Client::OPTION_ERROR_CODE, $user_id ) );
```

Comment:
I think this is because we aren't mocking the actual response of the OAuth request, but just returning a `200` response code.

I know we chatted about this and it's in-use elsewhere, and I think it's fine to consider fixing this out-of-scope of the issue because it's already in the codebase… but we should do two thing:

1. leave a more detailed explanation as to why this is happening and why we consider an invalid response/error code valid here (essentially, we're testing to make sure the request HAPPENED, not what it was)
2. file a follow-up issue that references the existing places we do this and this comment, to fix it

If you can attach that issue here that'd be great 😄

Reply:
I went down the route of mocking the OAuth Client (I've stashed those changes). However, I preferred the current solution as it did not require me to inject the OAuth client as a dependency into `Authentication.php`.

Anyways, I have added more comments and have created a new issue #4926 to fix this. While writing the description for this new issue, it kind of feels alright to do this probably due to the excessive clarity I have on this now. 😄